Kernel/Devices/AHCIPort.cpp

```diff
@@ -0,0 +1,438 @@
+{
+    setup_command(port);
+    m_dw0->cfl = sizeof(FIS_REG_H2D) / sizeof(u32);
+    m_dw0->write = 0;
+    *m_prdtl = 1;
+
+    m_entries[0].dba = (u32)out;
+    m_entries[0].dbau = 0;
+    m_entries[0].dbc = 256 * 2;
+    m_entries[0].i = 1;
+
+    memset((void*)m_cfis_reg_h2d, 0, sizeof(FIS_REG_H2D));
+    m_cfis_reg_h2d->fis_type = FIS_TYPE_REG_H2D;
+    m_cfis_reg_h2d->command = ATA_CMD_IDENTIFY;
+    m_cfis_reg_h2d->device = m_index;
+    m_cfis_reg_h2d->mode = 1;
+
+    return execute_command(port);
+}
+
+bool AHCISlot::issue_read(AHCIPort& port, u32 start, u16 count, u16* out)
+{
+    setup_command(port);
+    m_dw0->cfl = sizeof(FIS_REG_H2D) / sizeof(u32);
+    m_dw0->write = 0;
+    m_dw0->clear = 1;
+    m_dw0->prefetch = 0;
+    *m_prdtl = (u16)((count - 1) >> 4) + 1;
+
+    // 8K bytes (16 sectors) per table entry
+    int i, count_left = count;
+    for (i = 0; i < *m_prdtl - 1; i++) {
+        m_entries[i].dba = (u32)out;
+        m_entries[i].dbau = 0;
+        m_entries[i].dbc = 8 * 1024 - 1;
+        m_entries[i].i = 1;
+        out += 4 * 1024;
+        count_left -= 16;
+    }
+
+    // Make up for the remaining space with the last entry
+    m_entries[i].dba = (u32)out;
+    m_entries[i].dbau = 0;
+    m_entries[i].dbc = (count_left << 9) - 1;
+    m_entries[i].i = 1;
+
+    // Build command
+    memset((void*)m_cfis_reg_h2d, 0, sizeof(FIS_REG_H2D));
+    m_cfis_reg_h2d->fis_type = FIS_TYPE_REG_H2D;
+    m_cfis_reg_h2d->command = ATA_CMD_READ_DMA_EX;
+    m_cfis_reg_h2d->mode = 1;
+
+    m_cfis_reg_h2d->lba0 = (u8)(start >> 0) & 0xFF;
+    m_cfis_reg_h2d->lba1 = (u8)(start >> 8) & 0xFF;
+    m_cfis_reg_h2d->lba2 = (u8)(start >> 16) & 0xFF;
+    m_cfis_reg_h2d->lba3 = (u8)(start >> 24) & 0xFF;
+    m_cfis_reg_h2d->lba4 = (u8)(0);
+    m_cfis_reg_h2d->lba5 = (u8)(0);
+
+    m_cfis_reg_h2d->device = 1 << 6; // LBA mode
+    m_cfis_reg_h2d->countl = (count >> 0) & 0xFF;
+    m_cfis_reg_h2d->counth = (count >> 8) & 0xFF;
+
+    execute_command(port);
+    return true;
+}
+
+AHCIPort::AHCIPort(VirtualAddress addr, u32 index)
+    : m_base_addr(addr)
+    , m_index(index)
+{
+    // Assign data pointers
+    m_clb = (u32*)(m_base_addr.get() + AHCI_DEV_CLB);
+    m_clbu = (u32*)(m_base_addr.get() + AHCI_DEV_CLBU);
+    m_fb = (u32*)(m_base_addr.get() + AHCI_DEV_FB);
+    m_fbu = (u32*)(m_base_addr.get() + AHCI_DEV_FBU);
+    m_is = (u32*)(m_base_addr.get() + AHCI_DEV_IS);
+    m_ie = (u32*)(m_base_addr.get() + AHCI_DEV_IE);
+    m_cmd = (u32*)(m_base_addr.get() + AHCI_DEV_CMD);
+    m_tfd = (u32*)(m_base_addr.get() + AHCI_DEV_TFD);
+    m_sig = (u32*)(m_base_addr.get() + AHCI_DEV_SIG);
+    m_ssts = (u32*)(m_base_addr.get() + AHCI_DEV_SSTS);
+    m_serr = (u32*)(m_base_addr.get() + AHCI_DEV_SERR);
+    m_sact = (u32*)(m_base_addr.get() + AHCI_DEV_SACT);
+    m_ci = (u32*)(m_base_addr.get() + AHCI_DEV_CI);
+
+    // Allocate new memory region for rebase
+    m_ahci_base_region = MM.allocate_kernel_region(
```

Comment:
This doesn't look sane. Why we allocate `PAGE_ROUND_UP(1024 * 50)` range size in the virtual address space?

Reply:
Currently we provide a new allocation helper called `allocate_contiguous_kernel_region` in `MemoryManager`. Please use that instead. Otherwise there's no guarantee that you will get a physical contiguous region, which can be fatal when working with hardware. Also, I still don't understand why you want to allocate such a big range in memory.